### ISSUE TITLE
Ctrl-C does not interrupt borg check

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#7893]
Using borg 1.2.6. 
Running `borg check -p`.

Apart from printing '^C', Ctrl-C doesn't have an effect.

Pressing Ctrl-C twice didn't abort borg check.

_Originally posted by @rvanlaar in https://github.com/borgbackup/borg/issues/6057_

**Comment by ThomasWaldmann** [comment]
For `create`, borg behaves like this:

Pressing Ctrl-C once should make a checkpoint / commit the repo (that can take a while) and then abort.

Pressing Ctrl-C twice aborts immediately.

**Comment by ThomasWaldmann** [comment]
Reproducing:

## with remote (ssh://...) repo
- while doing the repository-part of check: interrupting with Ctrl-C is not possible (1, 2, any amount)
- while doing the archives-part of check: 2x Ctrl-C aborts the check

## with local repo
- while doing the repository-part of check: 2x Ctrl-C aborts the check
- while doing the archives-part of check: (not tested, but I assume that 2x Ctrl-C aborts the check)

**Comment by ThomasWaldmann** [comment]
Guess that the repository part of the check running on the remote side might cause this problem.

**Comment by ThomasWaldmann** [comment]
Hmm, this might be difficult to fix in borg 1.x:
- we ignore sigint for the ssh subprocess (because it must not be killed for clean borg shutdown)
- it is the ssh subprocess that outputs the remote `borg check --repository-only` output
- there is no output from the borg client process in this phase

**Comment by ThomasWaldmann** [comment]
Guess this likely has changed due to the sigint ctrl-c related fix a while ago:
- the fix was not to let Ctrl-C (SIGINT) kill the ssh subprocess, because the ssh connection is needed for clean shutdown of borg.
- after the fix, because ssh is not killed by Ctrl-C any more, there is no way to interrupt borg if there is no local terminal output of borg.

**Comment by ThomasWaldmann** [comment]
Workaround: shoot the ssh subprocess of borg in some other way, like e.g.: `kill <pid_of_ssh_subprocess`

**Comment by ThomasWaldmann** [comment]
Won't fix in borg 1.x. Workaround see above.

borg2 will work very differently, so this might be fixed in next beta.

**Comment by ThomasWaldmann** [comment]
@mr-raj12 try that if you work on borg check.

**Comment by mr-raj12** [comment]
> [@mr-raj12](https://github.com/mr-raj12) try that if you work on borg check.

<img width="1157" height="270" alt="Image" src="https://github.com/user-attachments/assets/11979204-9bce-424b-b089-19d6f7741cfd" />


Verified locally: ran borg check --repository-only and sent Ctrl-C twice
it aborts with exit code 130 instead of running to completion. 
Note: it takes 2× Ctrl-C, since check doesn't poll sig_int for a single-Ctrl-C clean stop.

**Comment by ThomasWaldmann** [comment]
OK. The idea of single ctrl-c is to stop soon, but cleanly (guess it does not matter for the non-repair read-only borg check, but would be good to have for --repair).

2x ctrl-c is rather "stop NOW!".

**Comment by mr-raj12** [comment]
This isn't finished yet. #9966 said "Part of #7893" on purpose, so there was no auto-close to fire.

It fixed the pack scan, `verify_data`, `rebuild_archives_directory` and `rebuild_archives`. Two phases that run earlier on `--repair` still don't poll `sig_int`:

- `build_chunkindex_from_repo()`, cache.py:841, pack loop at cache.py:905
- `ArchiveChecker.make_key()`, archive.py:1966

Both run before the check code is reached, so on a big repo the first Ctrl-C still looks like it does nothing, which is what this ticket is about. #10042 covers that.

This is milestoned 2.0.0b23, but #10042 is on b24, so it can't be finished in b23. Could someone with write access move the milestone? I don't have the permission.

**Comment by ThomasWaldmann** [comment]
OK, I changed the milestone.

**Comment by mr-raj12** [comment]
#10382 takes care of the one phase that was still left: `build_chunkindex_from_repo()` now polls `sig_int` in its pack loop and raises instead of handing back a partial index.

The other phase from my earlier comment, `ArchiveChecker.make_key()`, does not do a key recovery scan at all anymore — #10377 reduced it to `key_factory(repository)` — so this ticket can be closed once #10382 is merged.
